File: js/Neetcode_150/5_Sliding_Window/33_sliding_window_max.js
//Sliding Window Max: given an int arr 'nums' & an int k. return an arr which stores the max value in each window of size k in nums.

var maxSlidingWindow = function(nums, k) {
    const dq = []  // Monotonically Decreasing Deque to store indices for values of nums
    const ans = []
    for(let i=0; i<nums.length; i++){
        while(dq.length>0 && nums[dq[[dq.length-1]]] <= nums[i]) dq.pop() // if nums[i] > any prev element, then for next indexes(j) >= i, ans[j] != those prev elements. 
        dq.push(i)                      
        if(dq[0] < i+1-k) dq.shift()  // ie, if(window size = k+1, remove the first element) 
        if(i >= k-1) ans.push(nums[dq[0]]) //ie, if(window size >= k) push_ans;
    }
    return ans
};

File: c++/Neetcode_150/5_Sliding_Window/33_sliding_window_max.cpp
//Sliding Window Max: given an int arr 'nums' & an int k. return an arr which stores the max value in each window of size k in nums.

    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
       int n = nums.size();
       deque<int> dq;
       vector<int> ans;
       for(int i=0; i<n; i++){
            while(!dq.empty() && nums[dq.back()] <= nums[i]) dq.pop_back();
            dq.push_back(i);
            if(dq.front() < i+1-k) dq.pop_front();
            if(i >= k-1) ans.push_back(nums[dq.front()]);
       }
       return ans;
    }